apps/openchallenges/edam-etl/src/main.py

```diff
@@ -1,16 +1,100 @@
-"""Extract, transform and load EDAM concepts"""
-
+import pandas as pd
+import requests
 from os import getenv
+from typing import Optional
+import regex as re
 
 # Get config from the environment variables
-oc_db_url = getenv("OC_DB_URL")
-print(f"OC DB URL: {oc_db_url}")
 
-# TODO Download EDAM concepts from GitHub or S3 bucket (CSV file)
-print("[TODO] Download EDAM concepts from GitHub or S3 bucket (CSV file)")
+OC_DB_URL = getenv("OC_DB_URL")
+VERSION = getenv("VERSION")
+print(f"EDAM Version: {VERSION}")
+print(f"OC DB URL: {OC_DB_URL}")
+
+
+def download_edam_csv(url: str, version: str) -> Optional[bool]:
+    """Download EDAM concepts from GitHub or S3 bucket (CSV file)"""
+    print("Downloading the EDAM concepts from GitHub (CSV file)...")
+    try:
+        response = requests.get(url)
+        response.raise_for_status()  # Raise an exception for bad response
+        with open(f"EDAM_{version}.csv", "wb") as f:
+            f.write(response.content)
+        print("EDAM concepts downloaded successfully.")
+        return True
+    except requests.RequestException as e:
+        print(f"Error downloading EDAM concepts: {e}")
+        return None
+
+
+def transform_to_dataframe(version: str) -> pd.DataFrame | None:
+    """Transform the CSV to a DataFrame with indices starting from 1"""
+    print("Processing the EDAM concepts...")
+    try:
+        df = (
+            pd.read_csv(f"EDAM_{version}.csv", usecols=["Class ID", "Preferred Label"])
+            .rename(
+                columns={"Class ID": "class_id", "Preferred Label": "preferred_label"}
+            )
+            .reset_index(drop=True)
+            .reset_index(drop=False)
+            .rename(columns={"index": "id"})
+        )
+        print("EDAM concepts processed successfully.")
+        return df
+    except FileNotFoundError as e:
+        print(f"Error processing EDAM concepts: {e}")
+        return None
+
+
+def print_info_statistics(df: pd.DataFrame) -> None:
+    """Gather data about the EDAM ontology"""
+    if df is not None:
+        print(f"Number of Concepts Transformed: {len(df)}")
+        print(f"Column names: {df.columns.tolist()}")
+
+        # Define regex patterns for each concept
+        data_pattern = re.compile(r"data_", re.IGNORECASE)
+        operation_pattern = re.compile(r"operation_", re.IGNORECASE)
+        format_pattern = re.compile(r"format_", re.IGNORECASE)
+
+        # Initialize counts
+        data_count = 0
+        operation_count = 0
+        format_count = 0
+        other_count = 0
+
+        # Loop through the class_id column and count occurrences
+        for value in df["class_id"]:
+            if data_pattern.search(value):
+                data_count += 1
+            elif operation_pattern.search(value):
+                operation_count += 1
+            elif format_pattern.search(value):
+                format_count += 1
+            else:
+                other_count += 1
```

Comment:
Aside the missing concept, this implementation is OK.

However, "for" loops are usually inefficient. Could you please try the following implementation proposed by ChatGPT?

> Certainly! To implement a solution without an explicit for loop, we can leverage pandas' vectorized string operations. These operations allow us to apply regular expressions over the entire column efficiently. However, note that this approach may be more suitable for certain types of patterns and data structures. In this case, we'll use the str.contains() method provided by pandas, which supports regex and can be used to create boolean masks for filtering or counting.

```python
import pandas as pd

# Assuming df is your DataFrame and 'class_id' is the column of interest

# Create regex patterns for each concept
data_pattern = r"data_"
operation_pattern = r"operation_"
format_pattern = r"format_"

# Use pandas' vectorized string operations to count occurrences
data_count = df["class_id"].str.contains(data_pattern, case=False, na=False).sum()
operation_count = df["class_id"].str.contains(operation_pattern, case=False, na=False).sum()
format_count = df["class_id"].str.contains(format_pattern, case=False, na=False).sum()

# Calculate 'other' count by subtracting the specific counts from the total
other_count = len(df) - (data_count + operation_count + format_count)

# Example of how to print the counts
print(f"data: {data_count}")
print(f"operation: {operation_count}")
print(f"format: {format_count}")
print(f"other: {other_count}")
```

Reply:
The solution introduced by ChatGPT relies on `contains`. There is a way to use `contains` without bumping into the issue I described previously. Can you see how to achieve that?

Reply:
Note how this vectorized implementation is easier to read.

Reply:
The cconcepts were used in accordance to this comment:

![image](https://github.com/Sage-Bionetworks/sage-monorepo/assets/122999770/ae3c18bd-d859-4fe9-8d14-be2119221fb4)
I have added the Topic and Identifier concepts.

I avoided contains() due to this response to @vpchung comment:
![image](https://github.com/Sage-Bionetworks/sage-monorepo/assets/122999770/49549970-ed7f-483c-a37f-da43a9fcf488)
I have updated the function to include contains()

Reply:
`contains()` makes a comeback! :-)

Reply:
The contains() will not bump into an issue if the regex is appropriately written the way @vpchung suggested to include a hyphen after the word.

Reply:
Re: the comment about exact matches, this can be done with `contains()` by flagging `regex=True` which will allow searching with a pattern. I showed an usage example in my latest code review, but feel free to ping me for some guidance!

Reply:
@mdsage1 Do you mean the underscore character? A constraint should be placed on both sides of the string, not just at the end. For example (the forward slash may need to be escaped)

```
data_pattern = r"/data_"
```

Reply:
Sorry about that I thought I typed underscore.  I meant underscore not hyphen. I've added the regex=True (@vpchung) great catch because I thought I typed that in! I added a forward slash before the concept name @tschaffter